django/db/models/base.py

```diff
@@ -2457,7 +2458,9 @@ def _check_constraints(cls, databases):
 def method_set_order(self, ordered_obj, id_list, using=None):
     order_wrt = ordered_obj._meta.order_with_respect_to
     filter_args = order_wrt.get_forward_related_filter(self)
-    ordered_obj.objects.db_manager(using).filter(**filter_args).bulk_update(
+    ordered_obj.objects.db_manager(using or self._state.db).filter(
```

Comment:
Using `self._state.db` will prevent `db_for_write` fallback and crash if the object was retrieved from a read-only replica.

https://github.com/django/django/blob/e9fd2b572410b1236da0d3d0933014138d89f44e/django/db/models/query.py#L877-L880

https://github.com/django/django/blob/e9fd2b572410b1236da0d3d0933014138d89f44e/django/db/models/query.py#L1764-L1769

Reply:
Ah, absolutely ! I mixed up, this specific change is not needed. I reverted it in a new commit.

Reply:
I understand now why indeed this will break using routers.

However, for the other part of the fix (the filtering in `def _get_next_or_previous_in_order(self, is_next):`, I still believe it's good: the next and previous object *can only live in the same database as the current one*, no ?

Reply:
>  the next and previous object can only live in the same database as the current one, no ?

Yes but we have checks in place to ensure that it's the case at query evaluation time and both resulting `db_for_read` calls with have the exact same signature so (`router.db_for_read(self.__class__)`) so they should return the exact same `db`, why wouldn't they?